### PR TITLE
[READY]Implements Rever[B]!

### PR DESCRIPTION
**Pull request by Tacolizard**
:cl: Tacolizard
add: Sounds now have reverb effects, the reverb is determined by whichever area the sound listener is located.
add: you can toggle reverb with the 'Toggle reverb' button in the preferences tab
/:cl:

making the reverb based on the listener's location instead of the source may be less accurate, but in gameplay you won't really notice and it mostly solves a pretty big problem with sounds cutting out that would arise if the reverb was based on the source location.

**this pr was partially redesigned because the previous method this pr used to create reverb was hacky and over-complicated.**

**for mappers and coders:**
This reverb system is area-based. Each area can have a `sound_environment` var, which is meant to be set to one of the preset environment defines found in `code\_DEFINES\sound_presets.dm` This preset determines the sound of the reverb.

TODO:
[X] Make the reverb play again
[X] Implement basic area-based occlusion
[X] Test area-based occlusion
[X] Implement distance-based occlusion
[X] Implement line-of-sight based occlusion
[X] Test advanced occlusion
[X] Finish making reverb presets
~~[] Allow echo override so things like mechs can have custom reverb~~ I might add this in a later PR.

### CODE REVIEW

**Comment by Tacolizard** [comment]
WIP because i had it working using `environment` but porting it to `echo` has made none of the reverb effects work. It compiles and has no runtime errors, but the reverb doesn't play. I know it is possible because lummox created a tool that used `echo` to play environment presets.

**Comment by kevinz000** [comment]
why are you setting modes on everything from 644 to 755?
also no comment until i see this in action as i've never actually seen this before.

File: code/game/area/sound_environments_guide.dm [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+18	quarry
+19	plain
+20	parking lot
+21	sewer pipe
+22	underwater
+23	drugged
+24	dizzy
+25	psychotic
+
+For information on how to further customize reverb, view http://www.byond.com/docs/ref/info.html#/sound/var/echo
```
Comment (optimumtact) [inline code/game/area/sound_environments_guide.dm]: just include this comment in the other file lol how pointless

File: code/datums/sound_system.dm [hunk 2]
```diff
@@ -0,0 +1,34 @@
+/*
+This list is used because using the environment var (and its presets) changes the sound environment
+for ALL sounds, including sounds already playing. This means that if you play a sound without reverb, any
+sound that is already playing will have its reverb removed. The workaround is to use the echo var. Echo
+works on one sound only. However, echo does not have presets. This list stores the arguments to be passed
+to echo, in order to simulate an environment preset.
+*/
+/datum/sound/var/list/presets=list(\
+	list(7.5,   1.00,   -1000,  -100,   0,  1.49,  0.83, 1.0,  -2602, 0.007,  200,   0.011, 0.250,  0.00,  0.25, 0.000, -5.0,  5000.0),
```
Comment (ChangelingRain) [inline code/datums/sound_system.dm]: please add comments with what echo type these are to each of these, or maybe even merge it with the guide?

**Comment by Tacolizard** [comment]
@kevinz000 I assume you are referring to the areas. I did it so that they have reverb by default that somewhat matched what type of area they are.

Also, if you want to see badly implemented reverb, go to Goonstation.

File: code/game/area/sound_environments_guide.dm (hunk 1 above)
Comment (Tacolizard) [inline code/game/area/sound_environments_guide.dm]: this is from before i added the list file.

File: code/datums/sound_system.dm (hunk 2 above)
Comment (Tacolizard) [inline code/datums/sound_system.dm]: i merged it.

**Comment by ShizCalev** [comment]
Can you add the vars as defines in areas, that way it's a bit less of a constant reference between files thing when adding new areas? Something like ECHO_MOUNTAINS or ECHO_HANGAR just seems a bit more friendly imo.

**Comment by Tacolizard** [comment]
hopefully i replaced them all

**Comment by Tacolizard** [comment]
nope i missed one. ha

File: code/game/sound.dm [hunk 3]
```diff
@@ -42,6 +43,27 @@
...
 	if(isturf(turf_source))
 		var/turf/T = get_turf(src)
+		//var/area/hearer_location = get_area(T) //this will be used later
+		var/area/source_location = get_area(turf_source)
+
+		if(source_location != null && isarea(source_location))
+			var/area/A = source_location
+			if(A.sound_environment)
+				to_chat(world, "reverb triggered")
+				var/datum/sound/SD = new /datum/sound //todo: move this somewhere so it's only created once
```
Comment (MrPerson) [inline code/game/sound.dm]: Just make it static, boom done

**Comment by Tacolizard** [comment]
i never want to ctrl+c ctrl+v again

**Comment by Tacolizard** [comment]
this bug is driving me insane why can't i set echo to a define

**Comment by ShizCalev** [comment]
I'd break that out of misc.dm and into it's own file tbh

**Comment by Poojawa** [comment]
Are you really sure you want reverb?

https://www.youtube.com/watch?v=7kv55R5m-1w

**Comment by improvedname** [comment]
https://www.youtube.com/watch?v=9U2CePgS1Rc

**Comment by WJohn** [comment]
I swear to god if this causes echoes from fire alarms to be heard from anywhere on the station in a cacophony of "BEBEBEEPEPBEBEPEPEEP" I will find you and I will kill you.

**Comment by Repukan** [comment]
Can this stop exploding cats though?

**Comment by Tacolizard** [comment]
@WJohn I plan on implementing some pretty accurate occlusion to make that not happen.

**Comment by nicbn** [comment]
How will player-created space structures (aka autism forts) work?

**Comment by Tacolizard** [comment]
@nicbn I'm pretty sure adding an APC to an autism fort makes an area. If it doesn't, there is still a default environment set to the hangar preset, so the reverb will still work.

**Comment by ChangelingRain** [comment]
There should probably be a loc/turf override for echos, so you can have slightly different sound in mechs/lockers/on certain turfs/etc.

**Comment by Tacolizard** [comment]
@ChangelingRain That's a good idea, I'll add that to the todo list.

**Comment by Tacolizard** [comment]
Area based occlusion works, just needs some small tweaks

**Comment by RemieRichards** [comment]
@Tacolizard No.
APCs need Areas, they don't make them.
The little badge on your UI near the crafting and language menus makes areas for autism forts.

**Comment by Tacolizard** [comment]
@RemieRichards Well then it still will use the default reverb settings.

File: code/game/sound.dm [hunk 4]
```diff
@@ -110,15 +110,15 @@
 
 
 //this proc calls modify_echo with a preset used to create occlusion
-/proc/apply_occlusion(type, sin, distance)
+/proc/apply_occlusion(type, var/sound/sin, distance)
```
Comment (RemieRichards) [inline code/game/sound.dm]: no var/ in proc args, read the contributing guide ffs.

**Comment by Tacolizard** [comment]
I just realize we have code for finding line of sight, I'll use that to make even better occlusion

File: code/game/sound.dm [hunk 5]
```diff
@@ -33,6 +33,11 @@
 	S.wait = 0 //No queue
 	S.channel = channel || open_sound_channel()
 	S.volume = vol
+	S.environment = 10 //apparently you need to set an env for echo to work. dumb.
+	S.echo = ECHO_GENERIC //default echo, pretty much does nothing
+
+	var/debug_prints = TRUE //just for while this is WIP, it will be removed
```
Comment (Cyberboss) [inline code/game/sound.dm]: just use testing() reee

Comment (Tacolizard) [inline code/game/sound.dm]: wats testing()

why am i expected to know all these procs and vars that nobody documents or tells you about

Comment (Cyberboss) [inline code/game/sound.dm]: https://github.com/tgstation/tgstation/blob/master/code/_compile_options.dm#L3

**Comment by kevinz000** [comment]
@Tacolizard i meant you setting modes on the files themselves to 755

**Comment by optimumtact** [comment]
fix the permissions on the files damn it

File: code/game/sound.dm [hunk 6]
```diff
@@ -78,8 +106,41 @@
...
+			occlude_amount = 50
+		else
+			occlude_amount = 100
+
+		var/occlude = occlude_amount
+		var/distance
+		distance = get_dist(SA, LA)
+		occlude += (distance*occlude_amount) //each tile of distance = 100 more occlusion
+		var/list/modlist[18]
+		modlist[7] = occlude
```
Comment (optimumtact) [inline code/game/sound.dm]: I hope this is unfinished code here

Comment (Tacolizard) [inline code/game/sound.dm]: it's all unfinished code

**Comment by Tacolizard** [comment]
@kevinz000 I don't know how to fix that. I use DM in a VM, and commit changes using GitKraken on OSX.

**Comment by nicbn** [comment]
Stop using VM to ban evade and commit via GIT or upload bro

**Comment by KorPhaeron** [comment]
@tacolizard

Do you know if its possible to change the volume of a specific sound channel while sounds in it are ongoing?

**Comment by Tacolizard** [comment]
@KorPhaeron I know you can change the volume of single sounds while they're ongoing, I don't see any reason why it wouldn't work with channels, as channels also have a volume control. Of course in order to do that effectively you'd need to completely refactor the sound system because it's not designed to modify sounds as they are playing

**Comment by Tacolizard** [comment]
hello? anyone here? this PR is ready for review. It's not dead.

**Comment by BeeSting12** [comment]
Yeah, this is pretty bad. Makes every sound obnoxious to hear, even the small shit like putting something in a bag. I muted the game because of it.

**Comment by Tacolizard** [comment]
@BeeSting12 1. I will make the reverb quieter 2. I am going to add a preference to toggle it 3. admin sounds will not have reverb

**Comment by TrustyGun** [comment]
@Tacolizard 

using a preference for this is bad considering this actually impacts gameplay!

**Comment by Tacolizard** [comment]
@TrustyGun It doesn't impact gameplay. It doesn't make sounds easier or harder to hear because volume falloff is the same as it always was.

**Comment by Bawhoppen** [comment]
@TrustyGun we have preferences that impact gameplay anyways: see instruments

**Comment by Tacolizard** [comment]
uh, this PR is still here.

**Comment by optimumtact** [comment]
rip

**Comment by Tacolizard** [comment]
@optimumtact why rip

**Comment by Fox-McCloud** [comment]
As @RemieRichards and myself have pointed out a long time ago; this system doesn't actually work properly.

**Comment by optimumtact** [comment]
rip

**Comment by Fox-McCloud** [comment]
It basically has to do with how BYOND manipulates sound; it doesn't properly reset reset settings to it---as far as I know this has never been corrected; you'll end up messing up proper reverb by going with this.

**Comment by RemieRichards** [comment]
@Fox-McCloud I did say in IRC, and I even mentioned how it does mysteriously work for Bay, but he wouldn't take no for an answer.

**Comment by optimumtact** [comment]
it does work

**Comment by RemieRichards** [comment]
s/s/sn't

**Comment by Tacolizard** [comment]
>I even mentioned how it does mysteriously work for Bay

that doesn't make any sense. have you even tested this PR locally? What do you mean it 'messes up proper reverb'? I have thouroughly tested this PR and it works exactly as intended.

**Comment by Fox-McCloud** [comment]
You have to abuse a very particular process of constantly resetting the sound environment to get it to reset the modifications to the sound itself, otherwise, they end up being applied globally.

So it "works" on bay, but it and of itself is reliant on a poorly implemented system that works around the BYOND bug; it's still broken.

As a side note, I'd check into Para's sound system if you'd like improvements and the whole "3D sounds" thing that echo a bit; It's based on Bay's system, but has some major optimization upates to it: I recall @MrStonedOne wanting to check it out/implement it, but he got pretty busy.

**Comment by Tacolizard** [comment]
@Fox-McCloud All I can say is that through all the testing I have done, both personally and when this was testmerged on basil, it worked 100% as intended. I had no issues related to changing reverb globally. Again, I strongly encourage you to checkout this PR and test it yourself.

**Comment by Tacolizard** [comment]
@Fox-McCloud @RemieRichards I think what you guys are getting at is how environment sets globally, so if a sound with reverb plays it will override the reverb of all other playing sounds? I have solved this problem by making environment set by the location of the listener, not the sound source, so a sound from another area can't override other sounds, because all the reverb is the same? does that make sense?

**Comment by Tacolizard** [comment]
changed the title to try to attract hip younguns to this old pr

**Comment by Tacolizard** [comment]
where are muh hip younguns

**Comment by Shadowlight213** [comment]
@Tacolizard Reverb is applied to ahelp bwoinks and sounds bad.

**Comment by Tacolizard** [comment]
@Shadowlight213 
unfortunately that is a little bit difficult to fix with our current sound manager, I'm waiting on oranges to hopefully do a rewrite.

if no rewrite comes, I'd like to have this merged so that i can work on a sound manager rewrite in a seperate PR. I don't want to add it to this one for the sake of atomization. Remember, if ahelp reverb really bothers you, you can disable reverb.

**Comment by optimumtact** [comment]
I wouldnt' wait for me, my motivation is all over the place

**Comment by Fox-McCloud** [comment]
>Reverb is applied to ahelp bwoinks and sounds bad.

And now you're starting to see what @RemieRichards and I were both getting at.

**Comment by Tacolizard** [comment]
While this PR might not look like much, it wasn't always so simple. I spent a lot of time researching and doing trial and error to get this where it is. I really didn't want to close it, but I will.

**Comment by Tacolizard** [comment]
RIP

**Comment by p440** [comment]
wow rip

**Pull request closed without merging**